### ISSUE TITLE
Wrong steps per epoch while DDP training

### ISSUE THREAD

**Issue by xyangk** [Lightning-AI/lightning#13389]
## 🐛 Bug

I'm using DPP to train a encoder model with 5 GPU nodes, batch_size is 400 pre node,  model will gather all embeddings from 5 nodes to create a 2000*2000 matrix, then compute loss with this matrix.
Each node keeps 2623678 train samples and 491939 val samples, so the steps per epoch shoud be 2623678//400+491939//400 is 7788, BUT, the preocess bar show the steps per epoch is 1556, seems come from 2623678//2000+491939//2000
![image](https://user-images.githubusercontent.com/9495054/175318995-cdb980a9-24d7-41b5-af8e-f53c60645328.png)

trainer config：
```
trainer = pl.Trainer(accelerator='gpu', devices=1, auto_select_gpus=True, max_epochs=100,
                     callbacks=callbacks, precision=16,
                     log_every_n_steps=20, strategy="ddp_sharded", num_nodes=5)
trainer.fit(model=self.pl_model, datamodule=self.data_module)
```
The train dataloader and val dataloader created with batch size 400:
![image](https://user-images.githubusercontent.com/9495054/175320251-f830a4c4-cabe-4de2-b09b-eed1d0b15745.png)

### Environment

- PyTorch Lightning Version : 1.5.10
- PyTorch Version : 1.9.0+cu102
- Python version : 3.6.8
- OS : Linux
- CUDA/cuDNN version: 11.0
- How you installed PyTorch (`conda`, `pip`, source): pip

### Additional context

cc @justusschock @kaushikb11 @awaelchli @akihironitta @rohitgr7

**Comment by xyangk** [comment]
> Hello
> 
> Since you are using ddp with 5 GPUs, your dataset gets sharded across the 5 GPUs and therefore each GPU only sees 1/5 of the dataset.
> 
> You computed 7788 for the total batches across all GPUs (which is correct 😃), and if you divide that by 5 you get the 1556 shown in the progress bar.
> 
> **You need to imagine that there are 5 progress bars all running in parallel up to 1556, just only the first bar of the first process is visible in the terminal!**

Yeah, but actually total data size is 15578085, 1/5 of the total dataset is 3115617 (2623678+491939),  total batches is 15578085//400 (which is 38945),  so 7788 shoud be steps per epoch on every node.

**Comment by awaelchli** [comment]
> I splited my dataset to 5 parts, and put them on 5 nodes respectively,

Oh this is a very important detail!!!
In this case, make sure to tell Lightning not to add a distributed sampler, as explained by @rohitgr7.
You can set `Trainer(replace_sampler_ddp=False)`.

Because of the distributed sampler in PyTorch, you wouldn't need to split the dataset manually.

**Comment by xyangk** [comment]
> interesting... this is a new case.
> 
> I assume what's happening here is since you have kept different files on each node, no node is sharing the data at all but a dataloader is created with a different dataset on each node and each one of them will be seeing part of that dataset in it's process since DistributedSampler is added to each of them.
> 
> do you have 5 nodes with 1 GPU each or 5 nodes with 5 GPU each?

 5 nodes with 1 GPU each

**Comment by xyangk** [comment]
> > I splited my dataset to 5 parts, and put them on 5 nodes respectively,
> 
> Oh this is a very important detail!!! In this case, make sure to tell Lightning not to add a distributed sampler, as explained by @rohitgr7. You can set `Trainer(replace_sampler_ddp=False)`.
> 
> Because of the distributed sampler in PyTorch, you wouldn't need to split the dataset manually normally, as this would normally be done online through sampling.

Thank you guys, it works! 👍 Seems I didn't get the point of DDP, especially distributed sampler.

**Comment by awaelchli** [comment]
Awesome! Glad this was a relatively easy fix :)
Let us know if there are further questions.
